### PR TITLE
BUG,ENH: Fix internal ``__array_wrap__`` for direct calls

### LINKED ISSUE #27783: BUG: ``gentype.__array_wrap__`` doesn't make much sense
### Describe the issue:

As mentioned in the document, `diagflat ` should return a two-dimensional array. But it does not return a two-dimensional array when the input is the `var`'s output.

### Reproduce the code example:

```python
import numpy as np
print(np.diagflat(np.var(np.array([1,2,3]))).shape) # output is ()
```

### Error message:

_No response_

### Python and NumPy Versions:

1.26.4
3.12.7 | packaged by Anaconda, Inc. | (main, Oct  4 2024, 13:27:36) [GCC 11.2.0]

### Runtime Environment:

[{'numpy_version': '1.26.4',
  'python': '3.12.7 | packaged by Anaconda, Inc. | (main, Oct  4 2024, '
            '13:27:36) [GCC 11.2.0]',
  'uname': uname_result(system='Linux', release='6.8.0-48-generic', version='#48~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Mon Oct  7 11:24:13 UTC 2', machine='x86_64')},
 {'simd_extensions': {'baseline': ['SSE', 'SSE2', 'SSE3'],
                      'found': ['SSSE3',
                                'SSE41',
                                'POPCNT',
                                'SSE42',
                                'AVX',
                                'F16C',
                                'FMA3',
                                'AVX2'],
                      'not_found': ['AVX512F',
                                    'AVX512CD',
                                    'AVX512_KNL',
                                    'AVX512_KNM',
                                    'AVX512_SKX',
                                    'AVX512_CLX',
                                    'AVX512_CNL',
                                    'AVX512_ICL']}},
 {'architecture': 'Prescott',
  'filepath': '/path/lib/python3.12/site-packages/numpy.libs/libopenblas64_p-r0-0cf96a72.3.23.dev.so',
  'internal_api': 'openblas',
  'num_threads': 24,
  'prefix': 'libopenblas',
  'threading_layer': 'pthreads',
  'user_api': 'blas',
  'version': '0.3.23.dev'}]

### Context for the issue:

The result breaks the document spec.

### PR DESCRIPTION
**Pull request by seberg**
Since adding `return_scalar` as am argument, the array-wrap implementations were slightly wrong when that argument was actually passed and the function called directly.

NumPy itself rarely (or never) did so for our builtin types now so that was not a problem within NumPy.

Further, the scalar version was completely broken, converting to scalar even when such a conversion was impossible.

As explained in the code.  For array subclasses we NEVER want to convert to scalar by default.  The subclass must make that choice explicitly.
(There are plenty of tests for this behavior.)

Closes gh-27783

### CODE REVIEW

File: numpy/_core/src/multiarray/methods.c [hunk 1]
```diff
@@ -892,24 +892,24 @@ static PyObject *
...
+     * this function behaved in older NumPy versions.
+     */
+    int return_scalar = 0;
 
-    if (PyTuple_Size(args) < 1) {
-        PyErr_SetString(PyExc_TypeError,
-                        "only accepts 1 argument");
+    if (!PyArg_ParseTuple(args, "O!|OO&:__array_wrap__",
+                &PyArray_Type, &arr, &context_ignored,
+                &PyArray_OptionalBoolConverter, &return_scalar)) {
```
Comment (seberg) [inline numpy/_core/src/multiarray/methods.c:904]: I opted to just allow passing `None` for `return_scalar`.  I don't think anyone ever should, but it feels like it doesn't hurt (and we should have the default for when nothing is passed).

**Comment by seberg** [comment]
@mhvk maybe you can have a quick glance over.  I think it should be straight forward.

This should be back-portable, but not sure I think it is important enough to worry about it.
(Happy to delete the release note again, especially if we want to backport.)

File: numpy/_core/src/multiarray/methods.c (hunk 1 above)
Comment (mhvk) [reply numpy/_core/src/multiarray/methods.c:904]: Makes sense.

p.s. Had to check what `O!` did, but like that it removes the need for raising our own `TypeError`.

File: numpy/_core/src/multiarray/methods.c [hunk 2]
```diff
@@ -892,24 +892,24 @@ static PyObject *
 array_wraparray(PyArrayObject *self, PyObject *args)
 {
     PyArrayObject *arr;
-    PyObject *obj;
+    PyObject *context_ignored = NULL;
```
Comment (mhvk) [inline numpy/_core/src/multiarray/methods.c:895]: Had to do a slight double take of whether this was a flag for whether the context would be ignored.  How about just 
```
PyObject *UNUSED = NULL;  # for the context argument
```

File: numpy/_core/src/multiarray/methods.c [hunk 3]
```diff
@@ -892,24 +892,24 @@ static PyObject *
...
-    }
-    if (!PyArray_Check(obj)) {
-        PyErr_SetString(PyExc_TypeError,
-                        "can only be called with ndarray object");
-        return NULL;
-    }
-    arr = (PyArrayObject *)obj;
 
+    /*
+     * Subclasses must implement `__array_wrap__` with all the arguments.
```
Comment (mhvk) [inline numpy/_core/src/multiarray/methods.c:909]: I found this a bit hard to parse. How about
```
* If the array does not have our own type, we always wrap, even
* if `return_scalar` is set, so that we preserve the (presumably
* important) subclass information.
* If the type matches, we honor `return_scalar`, passing through
* PyArray_Return to convert any array with ndim=0 to scalar.
```

Though perhaps even better to just move the comment up, replacing the new comment on l.897?

File: numpy/_core/src/multiarray/scalartypes.c.src [hunk 4]
```diff
@@ -2035,29 +2036,33 @@ gentype_getarray(PyObject *scalar, PyObject *args)
     return ret;
 }
 
-static char doc_sc_wraparray[] = "sc.__array_wrap__(obj) return scalar from array";
+static char doc_sc_wraparray[] = "Array wrap to implementation for scalar types";
```
Comment (mhvk) [inline numpy/_core/src/multiarray/scalartypes.c.src:2039]: delete "to" - and maybe still spell out the dunder method for easier grepping, i.e., `"__array_wrap__ implementation for scalar types"`

File: numpy/_core/src/multiarray/scalartypes.c.src [hunk 5]
```diff
@@ -2035,29 +2036,33 @@ gentype_getarray(PyObject *scalar, PyObject *args)
...
 
+/*
+ * Array wrap for scalars, returning a scalar again preferentially.
```
Comment (mhvk) [inline numpy/_core/src/multiarray/scalartypes.c.src:2042]: Maybe
```
* __array_wrap__ for scalars, returning a scalar if possible.
```

File: numpy/_core/src/multiarray/scalartypes.c.src [hunk 6]
```diff
@@ -2035,29 +2036,33 @@ gentype_getarray(PyObject *scalar, PyObject *args)
...
+ * (note that NumPy itself may well never call this itself).
+ */
 static PyObject *
 gentype_wraparray(PyObject *NPY_UNUSED(scalar), PyObject *args)
 {
-    PyObject *obj;
     PyArrayObject *arr;
+    PyObject *context_ignored = NULL;
+    /* return_scalar should be passed, but it was a scalar so prefer scalar */
```
Comment (mhvk) [inline numpy/_core/src/multiarray/scalartypes.c.src:2050]:
```
/* return_scalar should be passed, but we're scalar, so we prefer scalars by default */
```

File: numpy/_core/src/multiarray/methods.c [hunk 7]
```diff
@@ -919,7 +919,11 @@ array_wraparray(PyArrayObject *self, PyObject *args)
                 PyArray_NDIM(arr),
                 PyArray_DIMS(arr),
                 PyArray_STRIDES(arr), PyArray_DATA(arr),
-                PyArray_FLAGS(arr), (PyObject *)self, obj);
+                PyArray_FLAGS(arr), (PyObject *)self, (PyObject *)arr);
+    }
+    else if (return_scalar) {
```
Comment (mhvk) [inline numpy/_core/src/multiarray/methods.c:924]: Absolute nitpick, but for similarity with the scalar case, perhaps
```
Py_INCREF(arr)
if (return_scalar) {
    ...
```
Since the previous clause had a return, we don't need an `else if` strictly.

Though arguably less clear, so feel free to ignore (really, just because I thought I saw  a difference in ref counting when reviewing).

File: numpy/_core/tests/test_arrayobject.py [hunk 8]
```diff
@@ -31,3 +31,39 @@ def test_matrix_transpose_equals_swapaxes(shape):
...
+def test_array_wrap(subclass):
+    # NumPy should allow `__array_wrap__` to be called on arrays, it's logic
+    # is designed in a way that:
+    #
+    # * Subclasses never return scalars by default (to preserve their
+    #   information).  They can choose to if they wish.
+    # * NumPy returns scalars, if `return_scalar` is passed as True to allow
+    #   manual calls to `arr.__array_wrap__` to do the right thing.
+
+    class MyArr(np.ndarray):
```
Comment (mhvk) [inline numpy/_core/tests/test_arrayobject.py:46]: Should one also test a subclass that does not override `__array_wrap__`? Or is that done elsewhere?

File: numpy/_core/tests/test_arrayobject.py [hunk 9]
```diff
@@ -31,3 +31,39 @@ def test_matrix_transpose_equals_swapaxes(shape):
...
+        def __array_wrap__(self, arr, context=None, return_scalar=None):
+            return super().__array_wrap__(arr, context, return_scalar)
+
+    arr = np.arange(3)
+    if subclass:
+        arr = arr.view(MyArr)
+
+    arr0d = np.array(3, dtype=np.int8)
+    # Third argument not passed, None, or True "decays" to scalar.
```
Comment (mhvk) [inline numpy/_core/tests/test_arrayobject.py:55]: Copy & paste error from the scalar test? For an array, only `True` causes decay to scalar. Maybe just
```
# With third argument True, ndarray allows "decay" to scalar.
```

File: numpy/_core/tests/test_arrayobject.py [hunk 10]
```diff
@@ -31,3 +31,39 @@ def test_matrix_transpose_equals_swapaxes(shape):
...
+
+    class MyArr(np.ndarray):
+        def __array_wrap__(self, arr, context=None, return_scalar=None):
+            return super().__array_wrap__(arr, context, return_scalar)
+
+    arr = np.arange(3)
+    if subclass:
+        arr = arr.view(MyArr)
+
+    arr0d = np.array(3, dtype=np.int8)
```
Comment (mhvk) [inline numpy/_core/tests/test_arrayobject.py:54]: What if `arr0d` is a subclass too? As written, I think that causes decay to a scalar (if `arr` is a subclass).

**Review by mhvk — COMMENTED** [review]
This looks good, and my comments are mostly about the phrasing of comments.

Only slightly larger item is whether there are already explicit tests for the case that `self` is a subclass and the item to be wrapped is the same subclass.

File: numpy/_core/tests/test_arrayobject.py (hunk 10 above)
Comment (seberg) [reply numpy/_core/tests/test_arrayobject.py:54]: True, but I don't think it should ever be a subclass!  I.e. that would defeat the purpose of wrapping (converting everything to plain arrays)?
(Technically, subclasses are accepted here.  But in practice, I think we must assume that the original subclass is meaningless, because we must return *our* subclass.)

Comment (seberg) [reply numpy/_core/tests/test_arrayobject.py:54]: Ah, sorry, I misunderstood.  While it is probably true that this shouldn't happen, I agree that I should anticipate it and fix the branch (return of scalar must only happens if `self` is a base-class array)

**Comment by seberg** [comment]
Thanks for the good spot, pushed a fix and hopefully made the comments/docs a bit clearer.  It is slightly annoying because we have to very strictly check for the 0-d case.

(I am ignoring that `scalar.__array_wrap__(subclass)` returns the subclass.  I think that is OK and strictly speaking I think the caller would be incorrect anyway.)

File: numpy/_core/src/multiarray/methods.c [hunk 11]
```diff
@@ -888,28 +888,39 @@ array_finalizearray(PyArrayObject *self, PyObject *obj)
...
+                &PyArray_Type, &arr, &UNUSED,
+                &PyArray_OptionalBoolConverter, &return_scalar)) {
         return NULL;
     }
-    if (!PyArray_Check(obj)) {
-        PyErr_SetString(PyExc_TypeError,
-                        "can only be called with ndarray object");
-        return NULL;
+
+    if (return_scalar && Py_TYPE(self) == &PyArray_Type && PyArray_NDIM(arr) == 0) {
```
Comment (mhvk) [inline numpy/_core/src/multiarray/methods.c:915]: the `NDIM` part can in principle be removed - `PyArray_Return` does that already.  Maybe slightly better? (If going that route, adjust the comment to `# Return scalar if NDIM==0.`

**Review by mhvk — APPROVED** [review]
One very small comment left, but feel free to ignore.

File: numpy/_core/src/multiarray/methods.c (hunk 11 above)
Comment (seberg) [reply numpy/_core/src/multiarray/methods.c:915]: Yeah, that is what I had... But then my test fails, because I somewhat prefer returning an `ndarray` rather than a subclass, and `PyArray_Return` won't do that.

**Comment by seberg** [comment]
I think I'll put this in then, thanks for the review @mhvk and spotting the subclass issue!.
Happy to follow-up if necessary.

File: numpy/_core/src/multiarray/methods.c (hunk 11 above)
Comment (mhvk) [reply numpy/_core/src/multiarray/methods.c:915]: > Yeah, that is what I had... But then my test fails, because I somewhat prefer returning an `ndarray` rather than a subclass, and `PyArray_Return` won't do that.

OK, makes sense now!
